src/IO/parseDateTimeBestEffort.cpp

```diff
@@ -562,8 +562,17 @@ ReturnType parseDateTime64BestEffortImpl(DateTime64 & res, UInt32 scale, ReadBuf
 {
     time_t whole;
     DateTimeSubsecondPart subsecond = {0, 0}; // needs to be explicitly initialized sine it could be missing from input string
-    if (!parseDateTimeBestEffortImpl<bool>(whole, in, local_time_zone, utc_time_zone, &subsecond))
-        return ReturnType(false);
+
+    if constexpr (std::is_same_v<ReturnType, bool>)
+    {
+        if (!parseDateTimeBestEffortImpl<bool>(whole, in, local_time_zone, utc_time_zone, &subsecond))
+            return ReturnType(false);
```

Comment:
Isn't simple `return false` enough here?

Reply:
Yep, nothing fancy required. That is a bool-specialization, we want just simple yes/no answer and there is no point in trying to parse sub-second part when parsing the main part has just failed.

Unfortunately, I couldn't find a more elegant way to do that without splitting into two compile-time branches.